### DISCUSSION TITLE
Tasmota connects to the external server of ESP WebTools

### DISCUSSION THREAD

**Discussion by wchulze** [arendst/Tasmota#17795] [Q&A] [answered]
In the Tasmoat source code I have found a connection to the external server of

**`ESP WebTools`**

in the file  **xdrv_62_improv.ino**

As I know, Tasmota should work independently of external web servers. I'm a bit surprised by this connection.
Can someone explain what's behind this implementation?

**Comment by barbudor** [comment, marked as answer]
Where do you see that this driver makes connection to  an external web site?
It is the implementation of a serial protocol that is used by the Tasmota web flasher

**Reply by sfromis** [reply]
More generally, while Tasmota does not require any external server connections, some features are based on the additional flexibility you can get this way, like getting firmware OTA updates, auto-configuration files and of course NTP. If you like, you can have local servers or avoid functions using external servers. There should be no "surprises" here, as long as you understand the basics.

**Comment by wchulze** [comment]
Thanks for the remark. I was amazed at the commented code like
`// Tasmota restarts after ssid or password change`
What this has to do with a web flasher is not clear to me. Furthermore, the driver is called up every second in normal operation, why.

**Reply by sfromis** [reply]
The restart is not up to the flasher, it is more like a reminder that Tasmota as such *will*  restart if you change ssid or password. Using `FUNC_EVERY_SECOND` is due to the driver needing to stay active during the setup, but will not do anything if not relevant.

**Reply by barbudor** [reply]
> What this has to do with a web flasher is not clear to me

Just try it : https://tasmota.github.io/install/
Makes the flashing and initial connection to your WLAN very easy

So yes the flasher itself, and the firmware are hosted on the Internet but you could just host that on your own local server at home if you want.

**Comment by wchulze** [comment]
Ok, there is a connection to the internet. I hope this is only active in case of programming. I think I will deactivate

`#define USE_IMPROV // Add support for IMPROV serial protocol as used by esp-web-tools (+2k code)`

I use an USB cable and an adapter on my desk. It couldn't be easier and faster from VSC !

**Reply by sfromis** [reply]
Well, there is no connection to the internet, unless you make the choice of using the web installer. The Improv driver cannot connect by being there, no such function in the code.

**Comment by wchulze** [comment]
Thanks for the info and the time spent.
